### DISCUSSION TITLE
Single-step debug problems using MPLAB X

### DISCUSSION THREAD

**Discussion by kjellc** [lcgamboa/picsimlab#124] [Q&A] [answered]
I really like Picsimlab and the way it makes it possible to test new SW and HW before it's built. 
I have tried to debug via MPLAB X, but it does not work all the way.
MPLAB-X (v6.20) can connect to Picsimlab, download my code and start execution.
But when execution is paused, or hits a breakpoint, the single step buttons (visible for a short while) disappear (!)
(F6 and F7 does not work either)

So it's impossible to single-step and/or continue execution.
What could be the problem? 

- I have tried several different boards and microcontrollers (only tried with PIC devices).
- I have added the "configuration bits" for my microcontroller to my C code.
- I have opened up the windows firewall to TCP and UDP ports 1234.
- I have installed the plugin "com-picsim-picsimlab.nbm" (tried different versions, latest, 0.9.1, 0.9.0)
- I have tried with both Windows and Linux (wsl2) versions of Picsimlab.
- I have tried different branches, latest, 0.9.1 and 0.9.0.
- I'm running MPLABX v6.20 on windows 11

**Comment by lcgamboa** [comment, marked as answer]
Hi @kjellc ,

 Microchip has changed the plugin's API but don´t update the [mplabx-sdk documentation](http://www.opensource4pic.org/content/content/mplab-x-sdk-mplab-x-ide) (again). So I have no information on how to update the plugin. The only way is to wait for the documentation to be updated and use older versions of MPLAB-X (6.05).

**Comment by kjellc** [comment]
Hi [lcgamboa](https://github.com/lcgamboa),

Thanks for your quick reply. Too bad Microchip keep changing the debug protocol API. :-(
I will revert to MPLABX v6.05 and try again.
And thanks for a great product!

**Comment by kjellc** [comment]
Single step works fine with MPLAB X v6.05. 
Thanks for your help!

**Comment by CycloComplex** [comment]
@lcgamboa can you message me your email address so we discuss further and provide you what you need to support the Simulator under MPLAB X IDE. Regards, Amit

**Comment by lcgamboa** [comment]
Hi @CycloComplex ,

my email is in all source files, lcgamboa@yahoo.com. 
The problem is that the https://github.com/lcgamboa/picsimlab_md plugin compiles with MPLABX6.20 but doesn´t work correctly. 
The SDK 6.20 hasn´t changed, but the behavior of MPLABX 6.20 has changed compared with previous versions.
With MPLABX6.20, debugging the plugin in NetBeans 8.2 is impossible because errors do not occur with MPLABX6.05.

**Comment by CycloComplex** [comment]
Hi Lcgamboa,

Thanks for connecting.
Can you let me know what support you need from us (Microchip) to get the tool (https://github.com/lcgamboa/picsimlab) working under the latest MPLAB X IDE?
Since we are not maintaining the tool, we won’t be able to fix the integration issues, but we can provide you with information or support so you can fix this issue.

Thanks and regards,
Amit

**Comment by lcgamboa** [comment]
@kjellc, I discovered the problem, the GetPreviousPC() method which was not implemented in MPLABX, started to be used in version 6.20. Since there is no documentation in the SDK and debugging does not work with MPLABX 6.20, I simply copied the code of the GetPC() method and it apparently worked again. I am waiting for the return of contact with Microchip about the documentation of the method and how to make the debugging of the plugin work. Below is the version of the plugin with the temporary fix.


[com-picsim-picsimlab.nbm.zip](https://github.com/user-attachments/files/18037960/com-picsim-picsimlab.nbm.zip)

**Comment by CycloComplex** [comment]
Hi,

I have sent your analysis report to the engineering team. I will update you once I hear back from them.

Regards
Amit

________________________________

**Comment by lcgamboa** [comment]
The problem has been fixed in commit https://github.com/lcgamboa/picsimlab_md/commit/9a52d2880c7696d81529b39073fe820326af3636 . Thanks to @CycloComplex  for the support and information to fix the problem.
